### ISSUE TITLE
echo_client, echo_server

### ISSUE THREAD

**Issue by park-jaehong** [litespeedtech/lsquic#85]
Hi, I succeeded compiling in ubuntu 16.04, tried to test echo_client and echo_server but got an error
I read EXAMPLES.txt but failed to solve 
I want to know how to use

![capture](https://user-images.githubusercontent.com/56724621/67091895-d46b0180-f1e8-11e9-9cb7-de52fec3e77c.PNG)

**Comment by dtikhonov** [comment]
```
./echo_server -c $hostname,chain.pem,privkey.pem
```

and

```
./echo_client -H $hostname -s 127.0.0.1:12345
```

**Comment by linxiaocong** [comment]
Hi Dmitri,

It seems handshake always fail if client disconnect and connect again. For example,

```
./echo_server -c 127.0.0.1,chain.pem,privkey.pem
```

and

```
./echo_client -H 127.0.0.1 -L debug // <- this will succeed

(Press Ctrl+C)

./echo_client -H 127.0.0.1 -s 127.0.0.1:12345 -L debug // <- this will fail with handshake: could not parse peer transport parameters (105 bytes)
```

**Comment by dtikhonov** [comment]
This form:

```
./echo_client -H 127.0.0.1 -L debug
```

does not work for me at all.  When I use valid `$hostname` both for `-c` argument for server and `-H` argument for client, it works fine.

**Comment by linxiaocong** [comment]
I have a testing cert which has a sni "127.0.0.1". Anyway I try with a real certificate with a real sni, sometimes hankshake still fail, with the server saying ```handshake: cannot generate transport parameters: 55```

And if I manually set the QUIC version to ```LSQVER_046```, everything works fine. It seems there issues with ietf quic.

Here is some debug information. Hope it will be useful.

```
Process 31491 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
    frame #0: 0x00000001000b89f7 quic-tunnel`lsquic_tp_encode(params=0x00007ffeefbf6598, buf="�, bufsz=224) at lsquic_trans_params.c:138:9
   135
   136 	    if (need > bufsz || need > UINT16_MAX)
   137 	    {
-> 138 	        errno = ENOBUFS;
   139 	        return -1;
   140 	    }
   141
Target 0: (quic-tunnel) stopped.
(lldb) p need
(size_t) $0 = 309
(lldb) p bufsz
(size_t) $1 = 224
(lldb) p params
(const transport_params *) $2 = 0x00007ffeefbf6598
(lldb) p *params
(const transport_params) $3 = {
  tp_flags = 19
  tp_numerics_u = {
    s = {
      init_max_stream_data_bidi_local = 0
      init_max_stream_data_bidi_remote = 1048576
      init_max_stream_data_uni = 12288
      init_max_data = 1572864
      idle_timeout = 30000
      init_max_streams_bidi = 100
      init_max_streams_uni = 3
      max_packet_size = 1370
      ack_delay_exponent = 3
      max_ack_delay = 25
      active_connection_id_limit = 7
    }
    a = ([0] = 0, [1] = 1048576, [2] = 12288, [3] = 1572864, [4] = 30000, [5] = 100, [6] = 3, [7] = 1370, [8] = 3, [9] = 25)
  }
  tp_disable_active_migration = '\0'
  tp_stateless_reset_token = {
    [0] = '�'
    [1] = 'K'
    [2] = '�'
    [3] = '�'
    [4] = 'C'
    [5] = 'Q'
    [6] = '�'
    [7] = 'c'
    [8] = '\x1c'
    [9] = '�'
    [10] = ':'
    [11] = '\x8c'
    [12] = 'F'
    [13] = '�'
    [14] = '2'
    [15] = '�'
  }
  tp_preferred_address = {
    ipv4_addr = ([0] = '\0', [1] = '\0', [2] = '\0', [3] = '\0')
    ipv4_port = 0
    ipv6_addr = {
      [0] = '\0'
      [1] = '\0'
      [2] = '\0'
      [3] = '\0'
      [4] = '\0'
      [5] = '\0'
      [6] = '\0'
      [7] = '\0'
      [8] = '\0'
      [9] = '\0'
      [10] = '\0'
      [11] = '\0'
      [12] = '\0'
      [13] = '\0'
      [14] = '\0'
      [15] = '\0'
    }
    ipv6_port = 0
    cid = {
      len = '\0'
      u_cid = {
        buf = {
          [0] = '\0'
          [1] = '\0'
          [2] = '\0'
          [3] = '\0'
          [4] = '\0'
          [5] = '\0'
          [6] = '\0'
          [7] = '\0'
          [8] = '\0'
          [9] = '\0'
          [10] = '\0'
          [11] = '\0'
          [12] = '\0'
          [13] = '\0'
          [14] = '\0'
          [15] = '\0'
          [16] = '\0'
          [17] = '\0'
          [18] = '\0'
          [19] = '\0'
        }
        id = 0
      }
    }
    srst = {
      [0] = '\0'
      [1] = '\0'
      [2] = '\0'
      [3] = '\0'
      [4] = '\0'
      [5] = '\0'
      [6] = '\0'
      [7] = '\0'
      [8] = '\0'
      [9] = '\0'
      [10] = '\0'
      [11] = '\0'
      [12] = '\0'
      [13] = '\0'
      [14] = '\0'
      [15] = '\0'
    }
  }
  tp_original_cid = {
    len = '�'
    u_cid = {
      buf = {
        [0] = '\0'
        [1] = '\0'
        [2] = '\0'
        [3] = '\0'
        [4] = '\0'
        [5] = '\0'
        [6] = '\0'
        [7] = '\0'
        [8] = '\0'
        [9] = '\0'
        [10] = '\0'
        [11] = '\0'
        [12] = '\0'
        [13] = '\0'
        [14] = '\0'
        [15] = '\0'
        [16] = '\x7f'
        [17] = '\x03'
        [18] = '�'
        [19] = '�'
      }
      id = 0
    }
  }
}
(lldb)
```

**Comment by dtikhonov** [comment]
> And if I manually set the QUIC version to LSQVER_046, everything works fine. It seems there issues with ietf quic.

You can also pass `-o version=Q046` to the client.

> Here is some debug information. Hope it will be useful.

Thank you so much for the bug report!  A fix is in the works.

**Comment by dtikhonov** [comment]
The fix is in release 2.4.10 -- please give it a whirl.

**Comment by linxiaocong** [comment]
bug fixed!
